R/theme-base-colors.R

```diff
@@ -35,45 +26,43 @@
 #' be shades of gray. `bs_theme_base_colors` overrides these hard-coded values
 #' with colors interpolated between `bg` and `fg`.
 #'
-#' @seealso [bs_theme_add_variables()]
+#' @family customizations
 #'
 #' @examples
 #'
 #' bs_theme_new("4+3")
-#' bs_theme_base_colors(bg = "#000060", fg = "skyblue",
-#'   accent = "orange", secondary = "silver")
+#' bs_theme_base_colors(bg = "#000060", fg = "skyblue")
 #'
 #' # You can apply further customizations here if desired, e.g.:
-#' bs_theme_add_variables("success" = "#1D7732")
+#' bs_theme_accent_colors(primary = "orange", secondary = "silver")
 #'
 #' if (interactive()) {
-#'   bs_theme_preview()
+#'   bs_theme_preview(with_themer = FALSE)
 #' }
 #'
 #' @export
-bs_theme_base_colors <- function(bg = "#FFFFFF", fg = "#000000",
-  accent = NULL, secondary = NULL) {
-
-  theme <- bs_theme_get()
-  if (is.null(theme)) {
-    stop("No bootstraplib theme is active (did you forget to call bs_theme_new()?)")
+bs_theme_base_colors <- function(bg = "#FFFFFF", fg = "#000000") {
+  if (is.null(bg)) {
+    stop("`bg` argument must not be NULL")
   }
-
-  results <- if (any(c("4", "4+3") %in% theme_version())) {
-    bs4_theme_base_colors(bg, fg, accent, secondary)
-  } else if ("3" %in% theme_version()) {
-    bs3_theme_base_colors(bg, fg, accent, secondary)
-  } else {
-    stop("bs_theme_base_colors doesn't recognize the active version of Bootstrap")
+  if (is.null(fg)) {
+    stop("`fg` argument must not be NULL")
   }
-  bs_theme_add(results)
+
+  args <- list(bg = bg, fg = fg)
+  args <- validate_and_normalize_colors(args)
+
+  dispatch_theme_setter("bs_theme_base_colors", list(
+    "4+3" = bs4_theme_base_colors,
+    "4" = bs4_theme_base_colors,
+    "3" = bs3_theme_base_colors), args)
 }
 
-bs4_theme_base_colors <- function(bg, fg, accent, secondary) {
-  white <- htmltools:::parseCssColors(bg)
-  black <- htmltools:::parseCssColors(fg)
+bs4_theme_base_colors <- function(args) {
+  white <- args$bg
+  black <- args$fg
 
-  grays <- colorRamp(c(white, black), alpha = TRUE)(0:10/10)
+  grays <- grDevices::colorRamp(c(white, black), alpha = TRUE)(0:10/10)
```

Comment:
This could be for another PR if you want, but I think we should be using `scales::colour_ramp()`

Reply:
I'm going to hold off on this right now, it's not a drop-in replacement because it returns `#RRGGBB` results instead of a color matrix; we'd have to at least do `col2rgb` on the result.